src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,682 @@
+    }
+
+    ucp_request_set_callback(req, send.cb, cb);
+    return req + 1;
+}
+
+ucs_status_ptr_t ucp_am_send_nb(ucp_ep_h ep, uint16_t id,
+                                const void *payload, size_t count,
+                                uintptr_t datatype, ucp_send_callback_t cb,
+                                unsigned flags)
+{
+    ucs_status_t status;
+    ucs_status_ptr_t ret;
+    ucp_request_t *req;
+    size_t length;
+    
+    if (ucs_unlikely((flags != 0) && !(flags & UCP_AM_SEND_REPLY))) {
+        return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+    }
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    if (ucs_likely(!(flags & UCP_AM_SEND_REPLY)) && 
+        (ucs_likely(UCP_DT_IS_CONTIG(datatype)))) {
+        length = ucp_contig_dt_length(datatype, count);
+        
+        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
+            status = ucp_am_send_short(ep, id, payload, length);
+            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+                UCP_EP_STAT_TAG_OP(ep, EAGER);
+                ret = UCS_STATUS_PTR(status);
+                goto out;
+            }
+        }
+    }
+    
+    status = ucp_ep_resolve_dest_ep_ptr(ep, ep->am_lane);
+    if (status != UCS_OK) {
+        ret = UCS_STATUS_PTR(status);
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_am_send_req_init(req, ep, payload, datatype, count, flags, id);
+    if (flags & UCP_AM_SEND_REPLY) {
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.reply_proto);
+    } else {
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.proto);
+    }
+
+out:
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+    return ret;
+}
+
+
+static ucs_status_t
+ucp_am_handler_common(ucp_worker_h worker, void *hdr_end,
+                      size_t hdr_size, size_t args_length,
+                      ucp_ep_h reply_ep, uint16_t am_id, 
+                      uint16_t desc_flag, unsigned am_flags) {
+    ucp_recv_desc_t *desc   = NULL;
+    uint16_t recv_flags = 0;
+    ucs_status_t status;
+
+    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        recv_flags |= desc_flag;
+    }
+
+    status = ucp_recv_desc_init(worker, hdr_end, hdr_size + args_length,
+                                0, am_flags, 0,
+                                recv_flags, 0, &desc);
+    status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
+                                      desc + 1,
+                                      args_length,
+                                      reply_ep,
+                                      UCP_CB_PARAM_FLAG_DATA);
+
+    if ((status == UCS_OK) && !(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
```

Comment:
Yes, I can improve readability, but it will be nested if statements. I will comment this out and put in the other way and we can decided which to use.

Reply:
After cleanup `} else if (status != UCS_OK && status != UCS_INPROGRESS) {` branch and others:
```
if (am_flags & UCT_CB_PARAM_FLAG_DESC) {
    return status;
}

if (status == UCS_OK) {
    ucp_recv_desc_release(desc);
    return UCS_OK;
} else if (status == UCS_INPROGRESS) {
    return UCS_OK;
} 

return status;
```
WDYT? Hope, I didn't miss something...

Reply:
That looks good. Also lets us do an unlikely check.